### DISCUSSION TITLE
How can I get all blog tags data?

### DISCUSSION THREAD

**Discussion by WinmezzZ** [facebook/docusaurus#5856] [Q&A] [answered]
I want to collect all blog post tags data in my **custom component**. 

I searched all the documents and Google and no right answer for me. Then I saw the [BlogTagsListPage](https://github.com/facebook/docusaurus/blob/44b87c7db8e3f47445ed3400890e5b8a10f75d74/packages/docusaurus-theme-classic/src/theme/BlogTagsListPage/index.tsx?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L19) component used it from `props`, I had tried it but i found it just work for this component only. 

So is there any api to get it? or any other way to resolve this.

I need your help honest. Thanks for you guys!

**Comment by Josh-Cena** [comment, marked as answer]
There's no way to query blog tags from external pages, unfortunately. We are still pondering if it's worth adding to the global data, but since there's such a high demand (>5 people asking already) I would try to implement this as part of #4492.

One way you could do this is through enhancing the blog plugin, regenerate the tags data, and inject it to your custom page. (https://github.com/facebook/docusaurus/issues/4492#issuecomment-805041391)

**Reply by WinmezzZ** [reply]
looks difficult to config it, any complete demo code? 

And I don't need add a new page, just like use in `BlogPostPage` component. So how do I set the `path` and `component` fileds value?

**Reply by slorber** [reply]
> looks difficult to config it, any complete demo code?

Unfortunately no complete demo, and you'll have to hack yourself to figure things out for your specific use-case. This is not something we support out of the box and there is not a single perfect solution to solve this problem.

It's not clear to me what exactly you want to do. Which non-blog page should show which tags exactly, in which form? Please be specific and show screenshots/mockups if relevant, and we might be able to guide to you a solution.

**Reply by WinmezzZ** [reply]
@slorber 

Thanks for your anwser, according to your description and some other issues, I had wrote below code.

## src/components/custom-tags-list-page
```jsx
export default props => {
  console.log(props);
  return <h1>Tags</h1>;
};
```

## plugins/costom-tags-list-plugin.js
```js
const blogPluginExports = require('@docusaurus/plugin-content-blog');

const blogPlugin = blogPluginExports.default;

function blogPluginEnhanced(context, options) {
  const blogPluginInstance = blogPlugin(context, options);

  return {
    ...blogPluginInstance,

    async contentLoaded(...contentLoadedArgs) {
      await blogPluginInstance.contentLoaded(...contentLoadedArgs);

      const { actions, content } = contentLoadedArgs[0];
      const { createData, addRoute } = actions;
      const { blogTags } = content;

      if (Object.keys(blogTags).length > 0) {
        const customTagsPath = await createData('custom-tags.json', JSON.stringify(blogTags));

        addRoute({
          path: '/custom-tags',
          component: '/src/components/custom-tags-list-page',
          exact: true,
          modules: {
            tags: customTagsPath
          }
        });
      }
    }
  };
}

module.exports = {
  ...blogPluginExports,
  default: blogPluginEnhanced
};

```

## docusaurus.config.js
```diff
module.exports = {
 presets: [
    [
      '@docusaurus/preset-classic',
      ({
        ...
+       blog: false,
-         blog: {
-           blogTitle: '博客首页',
-           path: './blog',
-           routeBasePath: '/',
-           showReadingTime: true,
-           blogSidebarTitle: '最近更新',
-           blogSidebarCount: 10
-         },
        // ...
      })
    ]
  ],
    plugins: [
+   [
+     './plugins/custom-tags-list-plugin.js',
+     {
+       blogTitle: '博客首页',
+       path: './blog',
+       routeBasePath: '/',
+       showReadingTime: true,
+       blogSidebarTitle: '最近更新',
+       blogSidebarCount: 10
+     }
+   ]
+ ],
}
```

then open the browser path 'localhost:3000/costom-tag', i saw the `tags` props, it work will!

![image](https://user-images.githubusercontent.com/40154766/140270583-12fbbd06-b4d0-44c4-aa25-be17761f96ef.png).

It's really half done here~

However, like I said above, I will use `tags` in built in components like `BlogListPage`, like this:

![image](https://user-images.githubusercontent.com/40154766/140271084-25e20ac5-b5b1-42dd-80f8-e42286a901a4.png).

I had run the swizzle `BlogTagsListPage` command, and changed plugin code:

```js
addRoute({
  path: '/',
  component: '/src/theme/BlogListPage',
  exact: true,
  modules: {
    tags: customTagsPath
  }
});
```

Every thing broken > _ <

seems like some problem with plugin, because `BlogListPage` component accept `metadata, items, sidebar` props also. May I need resolve all of route about blog?

What should I do next? Thanks.

**Reply by Josh-Cena** [reply]
Yes, you do need to give `BlogListPage` all props that it needs. And that very likely means you would have to duplicate the code from the actual plugin.

---

@slorber The more I think about it the more firmly I believe that #4492 is a justified need and we shouldn't pretend like the user doesn't know what she's doing by propagating front matter to global data. Speaking specifically about blog, things like recent posts and tag lists are very likely to be used outside the plugin, and in the same way we make docs versions global data, we should do the same for blog data.

I'm going to look into the API to make certain front matter become global data as I work on the `createFrontMatter` API.

**Reply by WinmezzZ** [reply]
Yeah, I'll try it. It's a big work to do.

**Reply by slorber** [reply]
The first thing we should do is to create some shared blog data bundle (containing tags) that are passed to every blog page => this will make it more convenient to display tags in multiple blog pages, and will be required for https://github.com/facebook/docusaurus/issues/5461

---

Now, the question is how do we display tags on non-blog pages, like your homepage or docs?

I'm not a fan of the "frontmatter propagation" to global data. Frontmatter only contains a subset of the useful data to inject in a page and it's not the most generic solution. Plugin extension already allows users to add global data themselves, including frontmatter, so I'd prefer we work on an extension API rather than a frontmatter propagation API:

```jsx
function blogPluginEnhanced(context, options) {
  const blogPluginInstance = blogPlugin(context, options);

  return {
    ...blogPluginInstance,
    async contentLoaded(...contentLoadedArgs) {
      await blogPluginInstance.contentLoaded(...contentLoadedArgs);
      const { actions, content } = contentLoadedArgs[0];
      const { setGlobalData } = actions;
      const { blogTags } = content;
	  setGlobalData({blogTags});	
    }
  };
}
```

A more generic solution is something like Gatsby do: the homepage is able to declare which data it needs thanks to a query. In our case, we could do something similar (without GraphQL) and the homepage could say `I need bundle "blog:instanceId:blogTags"` => less flexible API (compared to Gatsby) but simpler and it would work

**Reply by WinmezzZ** [reply]
It's works very will and easy to me to use `setGlobalData` and `useGlobalData`, thanks every much.

If it possible for updates on this topic please metion to me.

**Reply by dsmmcken** [reply]
```
async function blogPluginEnhanced(context, options) {
  const blogPluginInstance = await blogPlugin(context, options);

  return {
    ...blogPluginInstance,
    async contentLoaded(...contentLoadedArgs) {
      await blogPluginInstance.contentLoaded(...contentLoadedArgs);
      const { actions, content } = contentLoadedArgs[0];
      const { setGlobalData } = actions;
      const { blogTags } = content;
	  setGlobalData({blogTags});	
    }
  };
  ```
  
I was using a similar solution, and between beta.14 -> beta.18 this broke. Tracked it down to now needing to `await blogPlugin.`

This was tricky to debug, after upgrading and attempting to run gave just an ambiguous `[ERROR] TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined` only from attempting to `npm start`

**Reply by marklundin** [reply]
This does not seem to work anymore, It throws `TypeError: The "paths[0]" argument must be of type string. Received undefined`
